modules/navier_stokes/src/fvkernels/WCNSFV2PMomentumDriftFlux.C

```diff
@@ -65,6 +65,7 @@ WCNSFV2PMomentumDriftFlux::computeStrongResidual(const bool populate_a_coeffs)
 {
   _normal = _face_info->normal();
   const auto state = determineState();
+  const auto fd_state = Moose ::oldState();
```

Comment:
`oldState` will return the previous timestep even in a steady simulation, so there is a chance you will get wrong results in a steady simulation. Would using the previous newton iterate be enough?

Reply:
This would fetch the previous nonlinear iterate:
```
Moose::StateArg(1, Moose::SolutionIterationType::Nonlinear)
```

Reply:
I've tried that but not working. Let me show you the error message. Maybe you have better solution than this

Reply:
Using the previous newton iterate would probably deteriorate the performance of Newton. It sounds like the drift issue is for transients, so maybe `fd_state` could be `oldState()` for transients and just `determineState()` for steady?

Reply:
I would also prefer a more descriptive variable name than `fd_state`; I'm not sure what `fd` means

Reply:
I am looking at the code and there might be another issue underneath. We're debugging it right now. But we should be able to use this in Steady too.

Reply:
fd is the fraction_dispersed fraction

Reply:
So if you use current state you get the no-big-enough-container error. If you increase the container size, you get this (by Mengnan):
```
[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: Argument out of range
[0]PETSC ERROR: New nonzero at (2,81) caused a malloc
Use MatSetOption(A, MAT_NEW_NONZERO_ALLOCATION_ERR, PETSC_FALSE) to turn off this check
[0]PETSC ERROR: WARNING! There are unused option(s) set! Could be the program crashed before usage or a spelling mistake, etc!
[0]PETSC ERROR:   Option left: name:-i value: channel-drift-flux-transient.i source: command line
[0]PETSC ERROR:   Option left: name:-ksp_converged_reason value: ::failed source: code
[0]PETSC ERROR:   Option left: name:-snes_converged_reason value: ::failed source: code
[0]PETSC ERROR: See https://petsc.org/release/faq/ for trouble shooting.
[0]PETSC ERROR: Petsc Release Version 3.20.3, unknown 
[0]PETSC ERROR: /Users/lim2/Research/MOOSE/framework/worktrees/mixture_model_patch/modules/navier_stokes/navier_stokes-debug on a  named FN618356 by lim2 Wed Jul 17 10:23:14 2024
[0]PETSC ERROR: Configure options --with-64-bit-indices --with-cxx-dialect=C++17 --with-debugging=no --with-fortran-bindings=0 --with-mpi=1 --with-openmp=1 --with-shared-libraries=1 --with-sowing=0 --download-fblaslapack=1 --download-hypre=1 --download-metis=1 --download-mumps=1 --download-ptscotch=1 --download-parmetis=1 --download-scalapack=1 --download-slepc=1 --download-strumpack=1 --download-superlu_dist=1 --with-hdf5-dir=${PREFIX} --with-make-np=8 --COPTFLAGS=-O3 --CXXOPTFLAGS=-O3 --FOPTFLAGS=-O3 --with-x=0 --with-ssl=0 --with-mpi-dir=/Users/lim2/miniforge/envs/moose AR=arm64-apple-darwin20.0.0-ar RANLIB=arm64-apple-darwin20.0.0-ranlib CFLAGS="-ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -isystem /Users/lim2/miniforge/envs/moose/include   -mcpu=apple-a12" CXXFLAGS="-ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -isystem ${PREFIX}/include  -mcpu=apple-a12" CPPFLAGS="-D_FORTIFY_SOURCE=2 -isystem /Users/lim2/miniforge/envs/moose/include -mmacosx-version-min=11.3" FFLAGS="-march=armv8.3-a -ftree-vectorize -fPIC -fno-stack-protector -O2 -pipe -isystem /Users/lim2/miniforge/envs/moose/include   -march=armv8.3-a" FCFLAGS="-march=armv8.3-a -ftree-vectorize -fPIC -fno-stack-protector -O2 -pipe -isystem /Users/lim2/miniforge/envs/moose/include   -march=armv8.3-a" LDFLAGS="-Wl,-pie -Wl,-headerpad_max_install_names -Wl,-dead_strip_dylibs -Wl,-rpath,/Users/lim2/miniforge/envs/moose/lib -L/Users/lim2/miniforge/envs/moose/lib" --prefix=/Users/lim2/miniforge/envs/moose
[0]PETSC ERROR: #1 MatSetValues_SeqAIJ() at /opt/civet0/build/conda/conda-bld/moose-petsc_1712871010290/work/src/mat/impls/aij/seq/aij.c:458
[0]PETSC ERROR: #2 MatSetValues() at /opt/civet0/build/conda/conda-bld/moose-petsc_1712871010290/work/src/mat/interface/matrix.c:1497
```

Reply:
So here is what I think we should do: 
1. If you want to keep current for better performance, you will most likely need to increase the sparsity pattern. I suspect this is most likely coming from velocity depending on fd and we do a two-term for fd and it will depend on the gradient of fd and so on. You can try to disable to two-term expansion for fd in the variable. `use_two_term_expansion=false` or something similar to check if it helps.
2. I think we should have an option within the kernel to lag fd by evaluating it at the pervious nonlinear iterate using the line I sent here.

Reply:
Thanks a lot! Let me try this.

Reply:
Only use `two_term_boundary_expansion parameter` not working here, it gives the same error. It seems two_term_boundary_expansion doesn't affect the internal elements.  But I think that's the issue, I can add something similar for internal elements. It should fix the problem

Reply:
Aaa, this is on an internal element which does not touch the boundary? In that case try increasing the sparsity pattern to include 3 layers:
you can do it similarly to this:
```
params.set<unsigned short>("ghost_layers") = 3;
```
in the input parameters of the object. If this does not help there will some variable chain issues where one gradient depends on the other gradient on a face and they form very long dependency chains.

Reply:
You can also directly set `ghost_layers` in the input file when toying around.

I do not understand why we did not have these sparsity pattern errors before?

Reply:
some of these inputs were run with the segregated solver

Reply:
We didn't have this issue before because the input file was changed too. Instead of `rho_mixture_var` which is an auxvariable now we use `phase2` which is an actual nonlinear variable.

Reply:
got ya